feat: ratelimits on market service adapters

amitojsingh366 opened the pull request:
this implements rate limits for coingecko, coincap and yearn market service adapters
closes #367

time spend working on this: 4hours

gomesalexandre reviewed (changes requested):
A few minor comments, also discussing on Discord about the tests and how they can be improved

amitojsingh366 commented on packages/market-service/src/utils/getRatelimitedAxios.ts:
```diff
@@ -0,0 +1,6 @@
+
+export const getRatelimitedAxios = (rps: number) => {
+  return rateLimit(axios.create(), { maxRequests: rps, perMilliseconds: 1000, maxRPS: rps })
+}
```
fixed

amitojsingh366 commented on packages/market-service/src/coingecko/coingecko.test.ts:
```diff
@@ -1,13 +1,16 @@
 
-jest.mock('axios')
+// 8 requests per second as per https://www.coingecko.com/en/api_terms section 4.2
+const maxRPS = 8
```
fixed

gomesalexandre commented on packages/market-service/src/coingecko/coingecko.test.ts:
```diff
@@ -93,6 +95,15 @@ describe('coingecko market service', () => {
       expect(Object.keys(result).length).toEqual(0)
     })
 
+    it('does not get rate limited', async () => {
```
Discussed on Discord, not sure about the value of this since we're testing against some scenario (rate-limiting) that would never happen in tests.
Proposed to implement some poor man's rate limiting of mocks (send 429 if within rate limit, or response if not limited, checked depending on the numbers of axios.get calls per 1000ms) in tests if it's doable within reasonable effort, but I'm fine without it.

In this case, I'd vouch for removing the tests from that diff, but let's wait on another reviewers before doing so.

amitojsingh366 replied:
I've redone the tests now

0xdef1cafe commented on packages/market-service/src/constants.ts:
```diff
@@ -0,0 +1,4 @@
+// 8 requests per second as per https://www.coingecko.com/en/api_terms section 4.2
+export const COINGECKO_MAX_RPS = 8
```
please check the docs again https://www.coingecko.com/en/api/documentation it's 50/min not 8/s

amitojsingh366 replied:
hmmm the docs say 50/min but api TOS says 8/s, I will use the 50/min here then as that is the lesser one

0xdef1cafe commented on packages/market-service/src/utils/getRatelimitedAxios.ts:
```diff
@@ -0,0 +1,6 @@
+import axios from 'axios'
+import rateLimit from 'axios-rate-limit'
+
+export const getRatelimitedAxios = (rps: number, perMilliseconds = 1000) => {
+  return rateLimit(axios.create(), { maxRequests: rps, perMilliseconds })
+}
```
```suggestion
export const getRateLimitedAxios = (maxRPS: number) => rateLimit(axios.create()).setMaxRPS(maxRPS)
```

please read the docs carefully https://github.com/aishek/axios-rate-limit#usage

```
// note maxRPS is a shorthand for perMilliseconds: 1000, and it takes precedence
// if specified both with maxRequests and perMilliseconds
```

amitojsingh366 replied:
yeah that's why I removed `maxRPS` in favor of more granular control using `maxRequests` and `perMilliseconds`

0xdef1cafe commented:
@amitojsingh366 as we chatted about before 1:1, and yesterday as specifically related to this PR, please take the time to slow down and not rush with your PRs. more 🐢 less 🐇

in your most recent commit i can see that you haven't run the tests locally as they're failing in CI, after discussing with gomes recently.

with respect to tests, please add a new unit test for the new rate limiting util. if we have tests and confidence the rate limiting works, the market service unit tests should be specific to testing market service functionality, not the integration of the rate limiting and the market service at the same time.

we have three market service adapters, coingecko, coincap, and yearn. how does this address yearn?

**before resubmitting this PR for review, please** 

* read the docs and implement the rate limiting correctly for coingecko
* look at the precedence the args take for the rate limiting library, and make any required changes accordingly
* implement unit tests specific to the rate limiting functionality you've added
* ensure that the unit tests are not slowed down in real time testing rate limiting, you will need to investigate mocking timers if necessary.
* explain the rationale as to why yearn is not included in this PR

0xdef1cafe commented:
as discussed, 2/3 of my requested changes have not been addressed.

this

* slows down the unit tests - you will need to mock timers, and
* does not address why yearn has not been included.

additionally the rate limiting logic, including variable naming, is confusing at first glance, and without looking closely, seems wrong.

please leave the PR in draft until these are fully addressed and the logic is clear, or close the PR and stop working on the bounty if you don't wish to proceed.

amitojsingh366 commented:
ok so there's a discovery I made while implementing ratelimiting for the yearn adapter. the yearn api endpoints we use do **NOT** have ratelimits, and so for this reason I have set the maximum requests per second to 500 requests per minute which should be hard to trigger and should also be good for yearn servers.

here's how I found out yearn api has no ratelimits:
- found out the api endpoints that we use, they are listed below:
    -  <img width="871" alt="Screen Shot 2022-02-15 at 3 31 29 AM" src="https://user-images.githubusercontent.com/35400192/153969975-ac69921a-5552-4f15-a6a8-d4191c307f0c.png">
    - `https://cache.yearn.finance/v1/chains/1/vaults/get`
    - `https://cache.yearn.finance/v1/chains/1/ironbank/tokens`
    - `https://cache.yearn.finance/v1/chains/1/tokens/supported`
    - `https://cache.yearn.finance/v1/chains/1/vaults/tokens`
- spammed 2 out of these 4 endpoints to see if I get rate limited *(only 2 due to time constraints, script I used to spam and results are attached below)*
     - script used:
         - [index.js.txt](https://github.com/shapeshift/lib/files/8064855/index.js.txt) *(was not able to upload `.js` file so I added `.txt` at the end)*
     - `https://cache.yearn.finance/v1/chains/1/ironbank/tokens`:
         - [ironbank-tokens.txt](https://github.com/shapeshift/lib/files/8064849/ironbank-tokens.txt) *(was not able to upload `.json` file so I added `.txt` at the end)*
     - `https://cache.yearn.finance/v1/chains/1/vaults/tokens`
         - [vaults-tokens.json.txt](https://github.com/shapeshift/lib/files/8064861/vaults-tokens.json.txt) *(was not able to upload `.json` file so I added `.txt` at the end)*

amitojsingh366 commented:
@0xdef1cafe 

- [x] slows down the unit tests - you will need to mock timers: I removed timers as a whole
- [x] does not address why yearn has not been included: Added yearn adapter support, but yearn api does not have rate limits

time spent working: 6+ hours

0xdef1cafe commented:
hi @amitojsingh366 - i just approved the CI run and your tests still fail. i also took the time to check locally, and they fail there too.

after the extensive detailed guidance i've given here, the multiple attempts from your side, and time i've spent explaining implementation details and reviewing this PR, i'm going to close it.

thank you for the attempted contribution but we can't accept this.

amitojsingh366 commented:
> hi @amitojsingh366 - i just approved the CI run and your tests still fail. i also took the time to check locally, and they fail there too.
> 
> after the extensive detailed guidance i've given here, the multiple attempts from your side, and time i've spent explaining implementation details and reviewing this PR, i'm going to close it.
> 
> thank you for the attempted contribution but we can't accept this.

the tests pass on my laptop ☹️

![image](https://user-images.githubusercontent.com/35400192/154338269-86609bea-5f09-4057-b70f-2afe8113e90f.png)
